File: en/java-spring-webflux/src/main/java/com/example/controller/NotificationController.java
package com.example.controller;

import org.springframework.web.bind.annotation.*;
import com.example.model.Notification;
import com.example.dto.Result;
import com.example.dto.PageResult;
import jakarta.validation.Valid;
import java.util.List;

@RestController
@RequestMapping("/api/notifications")
public class NotificationController {
    
    @GetMapping
    public Result<PageResult<Notification>> listNotifications(
            @RequestParam Long userId,
            @RequestParam(required = false) String type,
            @RequestParam(required = false) Integer isRead,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @PostMapping
    public Result<Notification> createNotification(@Valid @RequestBody CreateNotificationReq req) {
        return Result.success(new Notification());
    }
    
    @GetMapping("/{id}")
    public Result<Notification> getNotification(@PathVariable Long id) {
        return Result.success(new Notification());
    }
    
    @DeleteMapping("/{id}")
    public Result<Void> deleteNotification(@PathVariable Long id) {
        return Result.success(null);
    }
    
    @PutMapping("/{id}/read")
    public Result<Notification> markAsRead(@PathVariable Long id) {
        return Result.success(new Notification());
    }
    
    @PutMapping("/user/{userId}/read-all")
    public Result<Void> markAllAsRead(@PathVariable Long userId) {
        return Result.success(null);
    }
    
    @GetMapping("/user/{userId}/unread-count")
    public Result<Long> getUnreadCount(@PathVariable Long userId) {
        return Result.success(0L);
    }
    
    @DeleteMapping("/user/{userId}/read")
    public Result<Void> deleteReadNotifications(@PathVariable Long userId) {
        return Result.success(null);
    }
    
    @DeleteMapping("/user/{userId}/all")
    public Result<Void> deleteAllNotifications(@PathVariable Long userId) {
        return Result.success(null);
    }
    
    @GetMapping("/user/{userId}/recent")
    public Result<List<Notification>> getRecentNotifications(
            @PathVariable Long userId,
            @RequestParam(defaultValue = "10") Integer limit) {
        return Result.success(List.of());
    }
    
    @GetMapping("/types")
    public Result<List<NotificationTypeVO>> getNotificationTypes() {
        return Result.success(List.of());
    }
    
    @PostMapping("/batch")
    public Result<List<Notification>> batchCreate(@Valid @RequestBody List<CreateNotificationReq> reqs) {
        return Result.success(List.of());
    }
    
    @PostMapping("/broadcast")
    public Result<Void> broadcastNotification(@Valid @RequestBody BroadcastNotificationReq req) {
        return Result.success(null);
    }
    
    @GetMapping("/user/{userId}/by-type")
    public Result<PageResult<Notification>> getByType(
            @PathVariable Long userId,
            @RequestParam String type,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @PutMapping("/batch-read")
    public Result<Void> batchMarkAsRead(@RequestBody List<Long> ids) {
        return Result.success(null);
    }
    
    @DeleteMapping("/batch")
    public Result<Void> batchDelete(@RequestBody List<Long> ids) {
        return Result.success(null);
    }
    
    @GetMapping("/user/{userId}/settings")
    public Result<NotificationSettingsVO> getSettings(@PathVariable Long userId) {
        return Result.success(new NotificationSettingsVO());
    }
    
    @PutMapping("/user/{userId}/settings")
    public Result<NotificationSettingsVO> updateSettings(
            @PathVariable Long userId,
            @Valid @RequestBody UpdateNotificationSettingsReq req) {
        return Result.success(new NotificationSettingsVO());
    }
    
    @PostMapping("/send-email")
    public Result<Void> sendEmailNotification(@Valid @RequestBody EmailNotificationReq req) {
        return Result.success(null);
    }
    
    @PostMapping("/send-sms")
    public Result<Void> sendSmsNotification(@Valid @RequestBody SmsNotificationReq req) {
        return Result.success(null);
    }
}

class CreateNotificationReq {
    private Long userId;
    private String title;
    private String content;
    private String type;
    
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    public String getType() { return type; }
    public void setType(String type) { this.type = type; }
}

class NotificationTypeVO {
    private String code;
    private String name;
    private String description;
    
    public String getCode() { return code; }
    public void setCode(String code) { this.code = code; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
}

class BroadcastNotificationReq {
    private String title;
    private String content;
    private String type;
    private List<Long> userIds;
    
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    public String getType() { return type; }
    public void setType(String type) { this.type = type; }
    public List<Long> getUserIds() { return userIds; }
    public void setUserIds(List<Long> userIds) { this.userIds = userIds; }
}

class NotificationSettingsVO {
    private Long userId;
    private Boolean emailEnabled;
    private Boolean smsEnabled;
    private Boolean pushEnabled;
    private List<String> enabledTypes;
    
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public Boolean getEmailEnabled() { return emailEnabled; }
    public void setEmailEnabled(Boolean emailEnabled) { this.emailEnabled = emailEnabled; }
    public Boolean getSmsEnabled() { return smsEnabled; }
    public void setSmsEnabled(Boolean smsEnabled) { this.smsEnabled = smsEnabled; }
    public Boolean getPushEnabled() { return pushEnabled; }
    public void setPushEnabled(Boolean pushEnabled) { this.pushEnabled = pushEnabled; }
    public List<String> getEnabledTypes() { return enabledTypes; }
    public void setEnabledTypes(List<String> enabledTypes) { this.enabledTypes = enabledTypes; }
}

class UpdateNotificationSettingsReq {
    private Boolean emailEnabled;
    private Boolean smsEnabled;
    private Boolean pushEnabled;
    private List<String> enabledTypes;
    
    public Boolean getEmailEnabled() { return emailEnabled; }
    public void setEmailEnabled(Boolean emailEnabled) { this.emailEnabled = emailEnabled; }
    public Boolean getSmsEnabled() { return smsEnabled; }
    public void setSmsEnabled(Boolean smsEnabled) { this.smsEnabled = smsEnabled; }
    public Boolean getPushEnabled() { return pushEnabled; }
    public void setPushEnabled(Boolean pushEnabled) { this.pushEnabled = pushEnabled; }
    public List<String> getEnabledTypes() { return enabledTypes; }
    public void setEnabledTypes(List<String> enabledTypes) { this.enabledTypes = enabledTypes; }
}

class EmailNotificationReq {
    private Long userId;
    private String email;
    private String subject;
    private String content;
    
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    public String getSubject() { return subject; }
    public void setSubject(String subject) { this.subject = subject; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
}

class SmsNotificationReq {
    private Long userId;
    private String phone;
    private String content;
    
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
}

File: en/kotlin-springmvc/src/main/kotlin/com/example/controller/NotificationController.kt
package com.example.controller

import com.example.dto.PageResult
import com.example.dto.Result
import com.example.model.Notification
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/notifications")
class NotificationController {
    
    @GetMapping
    fun listNotifications(
        @RequestParam(required = false) userId: Long?,
        @RequestParam(required = false) type: String?,
        @RequestParam(required = false) isRead: Int?,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Notification>> {
        val result = PageResult<Notification>()
        return Result.success(result)
    }
    
    @PostMapping
    fun createNotification(@RequestBody req: CreateNotificationReq): Result<Notification> {
        val notification = Notification()
        return Result.success(notification)
    }
    
    @GetMapping("/{id}")
    fun getNotification(@PathVariable id: Long): Result<Notification> {
        val notification = Notification()
        return Result.success(notification)
    }
    
    @PutMapping("/{id}")
    fun updateNotification(@PathVariable id: Long, @RequestBody req: UpdateNotificationReq): Result<Notification> {
        val notification = Notification()
        return Result.success(notification)
    }
    
    @DeleteMapping("/{id}")
    fun deleteNotification(@PathVariable id: Long): Result<Void> {
        return Result.success(null)
    }
    
    @PutMapping("/{id}/read")
    fun markAsRead(@PathVariable id: Long): Result<Notification> {
        val notification = Notification()
        return Result.success(notification)
    }
    
    @PutMapping("/read-all")
    fun markAllAsRead(@RequestParam userId: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/user/{userId}")
    fun getUserNotifications(
        @PathVariable userId: Long,
        @RequestParam(required = false) type: String?,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Notification>> {
        val result = PageResult<Notification>()
        return Result.success(result)
    }
    
    @GetMapping("/user/{userId}/unread-count")
    fun getUnreadCount(@PathVariable userId: Long): Result<Long> {
        val count = 0L
        return Result.success(count)
    }
    
    @DeleteMapping("/user/{userId}")
    fun deleteAllUserNotifications(@PathVariable userId: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/user/{userId}/unread")
    fun getUnreadNotifications(
        @PathVariable userId: Long,
        @RequestParam(defaultValue = "10") limit: Int
    ): Result<List<Notification>> {
        val notifications = listOf<Notification>()
        return Result.success(notifications)
    }
    
    @GetMapping("/types")
    fun getNotificationTypes(): Result<List<NotificationTypeVO>> {
        val types = listOf<NotificationTypeVO>()
        return Result.success(types)
    }
    
    @PostMapping("/batch")
    fun batchCreate(@RequestBody reqs: List<CreateNotificationReq>): Result<List<Notification>> {
        val notifications = listOf<Notification>()
        return Result.success(notifications)
    }
    
    @PostMapping("/broadcast")
    fun broadcastNotification(@RequestBody req: BroadcastNotificationReq): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/settings/user/{userId}")
    fun getUserSettings(@PathVariable userId: Long): Result<NotificationSettingsVO> {
        val settings = NotificationSettingsVO()
        return Result.success(settings)
    }
    
    @PutMapping("/settings/user/{userId}")
    fun updateUserSettings(
        @PathVariable userId: Long,
        @RequestBody req: UpdateNotificationSettingsReq
    ): Result<NotificationSettingsVO> {
        val settings = NotificationSettingsVO()
        return Result.success(settings)
    }
    
    @GetMapping("/statistics")
    fun getStatistics(): Result<NotificationStatisticsVO> {
        val statistics = NotificationStatisticsVO()
        return Result.success(statistics)
    }
    
    @GetMapping("/user/{userId}/statistics")
    fun getUserStatistics(@PathVariable userId: Long): Result<UserNotificationStatisticsVO> {
        val statistics = UserNotificationStatisticsVO()
        return Result.success(statistics)
    }
    
    @PostMapping("/{id}/send")
    fun sendNotification(@PathVariable id: Long): Result<Void> {
        return Result.success(null)
    }
    
    @DeleteMapping("/batch")
    fun batchDelete(@RequestBody ids: List<Long>): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/export")
    fun exportNotifications(
        @RequestParam(required = false) userId: Long?,
        @RequestParam(required = false) startDate: String?,
        @RequestParam(required = false) endDate: String?
    ): Result<String> {
        val fileUrl = ""
        return Result.success(fileUrl)
    }
}

data class CreateNotificationReq(
    val userId: Long?,
    val type: String?,
    val title: String?,
    val content: String?,
    val link: String?
)

data class UpdateNotificationReq(
    val title: String?,
    val content: String?,
    val link: String?
)

data class NotificationTypeVO(
    val code: String? = null,
    val name: String? = null,
    val description: String? = null
)

data class BroadcastNotificationReq(
    val type: String?,
    val title: String?,
    val content: String?,
    val userIds: List<Long>?
)

data class NotificationSettingsVO(
    val userId: Long? = null,
    val emailEnabled: Boolean? = null,
    val smsEnabled: Boolean? = null,
    val pushEnabled: Boolean? = null,
    val enabledTypes: List<String>? = null
)

data class UpdateNotificationSettingsReq(
    val emailEnabled: Boolean?,
    val smsEnabled: Boolean?,
    val pushEnabled: Boolean?,
    val enabledTypes: List<String>?
)

data class NotificationStatisticsVO(
    val totalNotifications: Long? = null,
    val sentNotifications: Long? = null,
    val readNotifications: Long? = null,
    val unreadNotifications: Long? = null
)

data class UserNotificationStatisticsVO(
    val userId: Long? = null,
    val totalNotifications: Long? = null,
    val readNotifications: Long? = null,
    val unreadNotifications: Long? = null
)
